apps/back-end/src/routes.ts

```diff
@@ -0,0 +1,189 @@
+ *
+ * Implementation Note: this is not a class, as ts-rest will attempt to recurse over its
+ * members, and any non-router members seem to make it explode when it does that.
+ * Therefore, we use a closure to store the configuration options, out of harm's way.
+ */
+export function MykomapRouter(
+  opts: MykomapRouterConfig,
+): RouterImplementation<Contract> {
+  // Validation
+  if (opts?.mykomap?.dataRoot == undefined)
+    // deliberately loose check
+    throw new Error("mandatory dataRoot option is not defined");
+
+  if (!fs.existsSync(opts.mykomap.dataRoot))
+    throw new Error(
+      `the dataRoot plugin option is set but refers to a non-existing path: `+
+        `'${opts.mykomap.dataRoot}'.`,
+    );
+
+  // Concatenates the path components into an absolute file path
+  const filePath = (...components: string[]): string => {
+    const p = path.join(opts.mykomap.dataRoot ?? "", ...components) + ".json";
+    return p;
+  };
+
+  // Construct and return the implementation object
+  return {
+    async getDataset({ params: { datasetId }, request, reply }) {
+      // Validate the parameters some more
+
+      if (!validBase64(datasetId))
+        throw new TsRestResponseError(contract.getDataset, {
+          status: 401,
+          body: { message: `invalid datasetId: '${datasetId}'` },
+        });
+
+      if (!sendJson(request, reply, filePath("datasets", datasetId)))
+        throw new TsRestResponseError(contract.getDataset, {
+          status: 404,
+          body: { message: `unknown datasetId '${datasetId}'` },
+        });
+
+      return reply;
+    },
+
+    async searchDataset({
+      params: { datasetId },
+      query: { filter, text },
+      request,
+      reply,
+    }) {
+      if (!validBase64(datasetId))
+        throw new TsRestResponseError(contract.searchDataset, {
+          status: 401,
+          body: { message: `invalid datasetId: '${datasetId}'` },
+        });
+
+      const filter2 = filter ?? [];
+      const invalidUris = filter2.filter((uri) => !validQName(uri));
+      if (invalidUris.length > 0)
+        throw new TsRestResponseError(contract.searchDataset, {
+          status: 400,
+          body: { message: `invalid filters: ${invalidUris}` },
+        });
+
+      const components = ["datasets", datasetId, "search", ...filter2, "text"];
+      if (text !== undefined) components.push(encodeURIComponent(text));
+
+      if (!sendJson(request, reply, filePath(...components)))
+        return {
+          status: 200,
+          body: [],
+        };
+
+      return reply;
+    },
+
+    async getDatasetItem({
+      params: { datasetId, datasetItemId },
+      request,
+      reply,
+    }) {
+      if (!validBase64(datasetId))
+        throw new TsRestResponseError(contract.getDatasetItem, {
+          status: 401,
+          body: { message: `invalid datasetId: '${datasetId}'` },
+        });
+      if (datasetItemId < 0)
```

Comment:
This can probably be added to the Zod schema in `contract.ts` too.

Reply:
Ditto above.